Several QoL (quality of life) improvements

embeddedt opened the pull request:
## Proposed Changes

  - Change hover color to be distinct from selection
    - See commit message.
  - Explicitly state when no option is selected on voting page
    - so students know if their selection has not been processed server-side.
  - Send a browser notification when a new question starts
    - I often switch away from my browser window; this will send a notification that should be visible even if the browser isn't.
  - Fix a typo in the README

## Tests

Please describe the tests that you ran to verify your changes.

- [x] Verified that selection still shows the same color
- [x] Verified that hovering has significant contrast from not hovering
- [x] Ran a local poll and verified that a notification is sent when a new question starts
- [x] Ran a local poll and verified that "Selected Option: None" is shown initially, instead of "Selected Option:" 
- [x] Tested on Edge for Windows 11 and verified that notifications are displayed consistently, even the second time

embeddedt committed:
Change hover color to be distinct from selection

Students were getting confused and thinking they had selected an
option when they were really just hovering on it. Also, touching
the option results in it remaining "hovered" even after the finger
is long gone.

embeddedt committed:
Explicitly state when no option is selected on voting page

embeddedt committed:
Send a browser notification when a new question starts

This is done in addition to changing the title as was done previously.

embeddedt committed:
Fix a typo in the README

hiimchrislim commented on client/src/pages/VotePage.tsx:
```diff
@@ -59,6 +65,13 @@ export const VotePage = () => {
         setSelectionOption("");
         if (!isFocus) {
           document.title = questionStarted;
+          if (notifPermission === "granted") {
```
Not all students will necessarily want this option all the time, so I would at least this as an icon that students can toggle (on the vote page)

hiimchrislim commented on client/src/pages/VotePage.tsx:
```diff
@@ -15,6 +15,7 @@ export const VotePage = () => {
   const cookies = new Cookies();
   const [pollCode] = useState(cookies.get(pollCodeCookie));
   const [started, setStarted] = useState(false);
+  const [notifPermission, setNotifPermission] = useState("denied");
```
I would probably change this to use a boolean and rename the state variables/function for it

shubhbapna replied:
+1

embeddedt replied:
Changed as requested (to `hasAllowedNotif`).

hiimchrislim commented on client/src/pages/VotePage.tsx:
```diff
@@ -59,6 +65,13 @@ export const VotePage = () => {
         setSelectionOption("");
         if (!isFocus) {
           document.title = questionStarted;
+          if (notifPermission === "granted") {
```
At least from testing this, it doesn't look like a notification will trigger for every new question (tested this on Windows 11). It seems like it'll trigger once, the user would have to either click on the notification to acknowledge it or to clear their notifications for it to retrigger. Just dismissing the notification doesn't seem to make it trigger anymore.

What's the functionality that you're trying to achieve for this? (For a notification on every single new question when out of focus?)

embeddedt replied:
> it doesn't look like a notification will trigger for every new question (tested this on Windows 11). It seems like it'll trigger once, then the user would have to either click on the notification to acknowledge it or to clear their notifications for it to retrigger. Just dismissing the notification doesn't seem to make it trigger anymore.

I fired up a VM and was also able to reproduce this behavior. It should be fixed with the latest commit.

> What's the functionality that you're trying to achieve for this? (For a notification on every single new question when out of focus?)

Yes, exactly. Currently it seems that only the title is updated when a new question starts, which means the user needs to be looking at the browser to be aware that they should vote. I think a notification is a cleaner solution and is probably better for accessibility purposes too (since a screen reader could announce it).

Notably, if a user does not want to see the notifications, they can simply turn off the permission altogether in the URL bar.

![image](https://user-images.githubusercontent.com/42941056/194755610-5c8f5ee4-1e5c-45f2-9006-5f5bebb47ab7.png)

hiimchrislim commented on client/tailwind.config.js:
```diff
@@ -16,7 +16,8 @@ module.exports = {
       secondary: "#FFFFFF",
-      hover: "#00B5B5",
+      hover: "#0060de",
+      selected: "#00B5B5",
```
Minor: I would keep the `hover` as what it was before (#00B5B5) and change the `selected` to (#0171b7).

It'll be something like this in the end which keeps it a bit more consistent with the colors scheme we're using.

![image](https://user-images.githubusercontent.com/24628243/194735784-00fd364e-bcbb-4bb2-b6f2-6c7b09af78c4.png) - Top is `selected` and bottom is `hover`

embeddedt replied:
Thanks. Color choice is not my strong suit. :laughing:

hiimchrislim reviewed (changes requested):
Left a few comments to review.

embeddedt committed:
Revert hover color and choose a new selected color

embeddedt committed:
Refactor notification permission state

embeddedt committed:
Refactor page focus/notification system

* Use document.visibilityState to determine when the tab is not focused.
  This avoids a race condition where the tab is unfocused before the
  page has loaded, so the app can't detect it. Also, this lets us
  get rid of an event handler.
* Forcefully close the notification that was previously open, to improve
  reliability of seeing the new notification.

hiimchrislim commented:
Will be looking at this during the weekend. (A bit busy at work atm)

logonoff committed:
Merge branch 'main' into qol-improvements

huonggiangbui commented:
@hiimchrislim are you able to check this?

shubhbapna commented:
@embeddedt is it possible to open the PR against gdsc open source branch. I want to include this in the next release scheduled for first of week of winter

embeddedt commented:
I switched the base and the commits seem to still be intact; let me know if I still need to create a new PR.

shubhbapna reviewed (approved):
LGTM

hiimchrislim committed:
Minor changes

hiimchrislim committed:
Merge gdsc open source branch

hiimchrislim committed:
Revert some changed files
